core/shared/src/main/scala/scalaz/zio/Promise.scala

```diff
@@ -187,8 +187,8 @@ object Promise {
     } yield b
 
   private[zio] object internal {
-    sealed abstract class State[E, A]                             extends Serializable with Product
-    final case class Pending[E, A](joiners: List[Callback[E, A]]) extends State[E, A]
-    final case class Done[E, A](value: ExitResult[E, A])          extends State[E, A]
+    sealed abstract class State[E, A]                               extends Serializable with Product
+    final case class Pending[E, A](joiners: List[IO[E, A] => Unit]) extends State[E, A]
+    final case class Done[E, A](value: IO[E, A])                    extends State[E, A]
```

Comment:
Is there a scenario where `value` would be something else than `IO.Fail` or `IO.Strict`? I wonder if we should enforce that with a `PureIO` type or something.